### PR TITLE
Add support for draft-ietf-ntp-roughtime-10

### LINKED ISSUE #53: Implement draft-ietf-ntp-roughtime-10
The only substantive change since draft-08 is the new SRV tag, which we will want to implement:
https://author-tools.ietf.org/iddiff?url1=draft-ietf-ntp-roughtime-08&url2=draft-ietf-ntp-roughtime-10&difftype=--html

We could easily support draft 10 alongside draft 08 and Google Roughtime.

### PR DESCRIPTION
**Pull request by cjpatton**
Closes #53 

The only change in this draft is the SRV tag used to negotiate the root public key. Concretely, the client sends this tag to indicate which root public key it expects to use to verify the response; when the server receives this tag, it will use it to guide selection of the certificate to use in the response.

This change also moves the online private key into the `Certificate` structure - currently this is passed separately into functions that need both the certificate and the private key. This is intended to simplify multi-certificate setups. Accordingly, the `NewCertificate()` takes as input the online private key rather than the public key. We also use the types in the `ed25519` package, rather than `[]bytes`, to distinguish public from private keys.

Finally, encapsulate the parsed request into a type, `Request`.

### CODE REVIEW

File: protocol/protocol_test.go [hunk 1]
```diff
@@ -370,6 +371,11 @@ func TestIETFTags(t *testing.T) {
 		got  uint32
 		want uint32
 	}{
+		{
+			name: "SRV",
+			got:  tagSRV,
+			want: 0x00565253,
```
Comment (cjpatton) [inline protocol/protocol_test.go:377]: Reviewer note: The bytes in Table 2 are incorrect: https://github.com/wbl/roughtime-draft/pull/10

File: protocol/protocol_test.go [hunk 2]
```diff
@@ -429,3 +436,101 @@ func TestEmptyVersionPreference(t *testing.T) {
...
+				t.Fatal("no certificate selected")
+			}
+
+			if !bytes.Equal(selectedCert.srv, certs[i].srv) {
+				t.Fatalf("selected the wrong certificate")
+			}
+		})
+	}
+
+	t.Run("uknown_server", func(t *testing.T) {
```
Comment (lukevalenta) [inline protocol/protocol_test.go:473]:
```suggestion
	t.Run("unknown_server", func(t *testing.T) {
```

File: protocol/protocol_test.go [hunk 3]
```diff
@@ -429,3 +436,101 @@ func TestEmptyVersionPreference(t *testing.T) {
...
+		}
+
+		req, err := ParseRequest(reqBytes)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		selectedCert := SelectCertificateForRequest(req, certs)
+		if selectedCert != nil {
+			t.Fatal("certificated selected for unknown server")
```
Comment (lukevalenta) [inline protocol/protocol_test.go:488]:
```suggestion
			t.Fatal("certificate selected for unknown server")
```

File: protocol/protocol_test.go [hunk 4]
```diff
@@ -429,3 +436,101 @@ func TestEmptyVersionPreference(t *testing.T) {
...
+		}
+
+		req, err := ParseRequest(reqBytes)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		selectedCert := SelectCertificateForRequest(req, []Certificate{})
+		if selectedCert != nil {
+			t.Fatal("certificated selected fromm empty list")
```
Comment (lukevalenta) [inline protocol/protocol_test.go:507]:
```suggestion
			t.Fatal("certificate selected from empty list")
```

**Review by lukevalenta — APPROVED** [review]
Couple minor typos but lgtm otherwise!

File: protocol/protocol.go [hunk 5]
```diff
@@ -288,12 +289,22 @@ func decodeFramed(req []byte) ([]byte, bool, error) {
...
+	return false
+}
+
 // CreateRequest creates a Roughtime request given an entropy source and the
 // contents of a previous reply for chaining. If this request is the first of a
 // chain, prevReply can be empty. It returns the nonce (needed to verify the
 // reply), the blind (needed to prove correct chaining to an external party)
 // and the request itself.
-func CreateRequest(versionPreference []Version, rand io.Reader, prevReply []byte) (nonce, blind []byte, request []byte, err error) {
+func CreateRequest(versionPreference []Version, rand io.Reader, prevReply, rootPublicKey []byte) (nonce, blind []byte, request []byte, err error) {
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply protocol/protocol.go]: Done.

File: protocol/protocol.go [hunk 6]
```diff
@@ -288,12 +289,22 @@ func decodeFramed(req []byte) ([]byte, bool, error) {
...
+	return false
+}
+
 // CreateRequest creates a Roughtime request given an entropy source and the
 // contents of a previous reply for chaining. If this request is the first of a
 // chain, prevReply can be empty. It returns the nonce (needed to verify the
 // reply), the blind (needed to prove correct chaining to an external party)
 // and the request itself.
-func CreateRequest(versionPreference []Version, rand io.Reader, prevReply []byte) (nonce, blind []byte, request []byte, err error) {
+func CreateRequest(versionPreference []Version, rand io.Reader, prevReply, rootPublicKey ed25519.PublicKey) (nonce, blind []byte, request []byte, err error) {
```
Comment (lukevalenta) [inline protocol/protocol.go:307]: nit: `prevReply` isn't a `ed25519.PublicKey`
```suggestion
func CreateRequest(versionPreference []Version, rand io.Reader, prevReply []byte, rootPublicKey ed25519.PublicKey) (nonce, blind []byte, request []byte, err error) {
```

Comment (cjpatton) [reply protocol/protocol.go:307]: That was incredibly sloppy of me. I'll put up a PR to fix, one sec.

Comment (cjpatton) [reply protocol/protocol.go:307]: https://github.com/cloudflare/roughtime/pull/56

**Pull request merged**